### PR TITLE
BoxesAPI: Get boxes by address considering mempool #137

### LINKED ISSUE #137: BoxesAPI: Get boxes by address considering mempool

### PR DESCRIPTION
**Pull request by monyedavid**
Fixes #137

### CODE REVIEW

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+@derive(encoder, decoder)
+final case class OutputInfoM(filteredBoxes: Items[OutputInfo], mempoolBoxes: Items[UOutputInfo])
+
+object OutputInfoM {
+
+  implicit val schema: Schema[OutputInfoM] =
+    Schema
+      .derived[OutputInfoM]
+      .modify(_.filteredBoxes)(_.description("Filtered boxes not spent in mempool"))
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala:16]: What does "Filtered boxes not spent in mempool" mean?

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package org.ergoplatform.explorer.http.api.v1.models
+
+import org.ergoplatform.explorer.http.api.models.Items
+import derevo.circe.{decoder, encoder}
+import derevo.derive
+import sttp.tapir.{Schema, Validator}
+
+@derive(encoder, decoder)
+final case class OutputInfoM(filteredBoxes: Items[OutputInfo], mempoolBoxes: Items[UOutputInfo])
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala:9]: Why so weird response format is chosen? I think it's better to return union of all boxes (mempool+non-mempool) where each entry is wrapped into some kind of wrapper carrying `confirmed` (`confirmed=false` for mempool boxes) flag.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/defs/BoxesEndpointDefs.scala [hunk 3]
```diff
@@ -123,11 +123,11 @@ final class BoxesEndpointDefs[F[_]](settings: RequestsSettings) {
       .in(paging(settings.maxEntitiesPerRequest))
       .out(jsonBody[Items[OutputInfo]])
 
-  def getOutputsByAddressDef: Endpoint[(Address, Paging), ApiErr, Items[OutputInfo], Any] =
+  def getOutputsByAddressDef: Endpoint[(Address, Paging), ApiErr, OutputInfoM, Any] =
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/defs/BoxesEndpointDefs.scala:126]: We shouldn't break existing API format. Please add separate endpoint for that feature.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/routes/BoxesRoutes.scala [hunk 4]
```diff
@@ -112,7 +119,14 @@ final class BoxesRoutes[
 
   private def getOutputsByAddressR: HttpRoutes[F] =
     interpreter.toRoutes(defs.getOutputsByAddressDef) { case (address, paging) =>
-      service.getOutputsByAddress(address, paging).adaptThrowable.value
+      (for {
+        spentBoxIds    <- mempool.getUSpentOutputsByAddress(address).adaptThrowable
+        mempoolOutputs <- mempool.getUOutputsByAddress(address).adaptThrowable
+        filteredOutputs <-
+          service
+            .getFilteredOutputsByAddress(address, paging, NonEmptyList.ofInitLast(spentBoxIds, BoxId("")))
```
Comment (oskin1) [inline modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/routes/BoxesRoutes.scala:127]: What is this `BoxId("")`?

**Review by oskin1 — CHANGES_REQUESTED** [review]
Please add tests for case when an address has mempool+confirmed boxes.

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala (hunk 2 above)
Comment (monyedavid) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala]: I have done this;
used mainChain=false for Mempool boxes

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/defs/BoxesEndpointDefs.scala (hunk 3 above)
Comment (monyedavid) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/defs/BoxesEndpointDefs.scala]: done

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/routes/BoxesRoutes.scala (hunk 4 above)
Comment (monyedavid) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/routes/BoxesRoutes.scala]: updated this to use Option[NonEmptyList], from .toNel

File: modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala (hunk 1 above)
Comment (monyedavid) [reply modules/explorer-api/src/main/scala/org/ergoplatform/explorer/http/api/v1/models/OutputInfoM.scala]: unspent boxes with consideration of the mempool (i.e boxes that have been used in an unconfirmed "UOutput" are filtered out).
I have updated this to return union of all boxes (mempool+non-mempool)

**Pull request closed without merging**